### ISSUE TITLE
[bug] user info / user groups

### ISSUE THREAD

**Issue by jeanpommier** [geobretagne/mviewerstudio#52]
Hi,
The conf has 2 parameters I was experimenting with yesterday:
- "user_info": "srv/user_info.php",
- "user_info_visible": true,

If you set "user_info_visible" to true, it triggers a block of code in https://github.com/geobretagne/mviewerstudio/blob/master/js/mviewerstudio.js#L140 that fails (with the above config at least):
Line 148, it checks for the `data.userGroups` length. But the data object is returned by the `user_info` service. The default one doesn't contain a userGroups  object. Hence the javascript failure

> Uncaught TypeError: Cannot read property 'length' of undefined
    at Object.success (mviewerstudio.js:237)
    at i (jquery.min.js:2)
    at Object.fireWith [as resolveWith] (jquery.min.js:2)
    at y (jquery.min.js:4)
    at XMLHttpRequest.c (jquery.min.js:4)

**Comment by bchartier** [comment]
Sorry for this.
This info is provided by Neogeo mviewerstudio backend and should be optional.
We may add a test to check the presence of this data.userGroups before testing its length.

**Comment by jeanpommier** [comment]
Yep ;o)

**Comment by bchartier** [comment]
would you fix this or would I?

**Comment by jeanpommier** [comment]
As you wish. I can't today anyway (working on another project)

**Comment by bchartier** [comment]
I would prefer you fix it. The reason is not my lazyness but as I test with Neogeo backend it would be easier for you to check the fix.

**Comment by jeanpommier** [comment]
OK @lazyb ;o)

**Comment by jeanpommier** [comment]
Mmmh, @bchartier , @spelhate @RemiDesgrange  it would be great to agree on the content of this user info. srv/user_info.php returns 
```
{"first_name": "Jean", "last_name":"Pommier", "organisation": {"legal_name":""}}
```
Given the [currrent code](https://github.com/geobretagne/mviewerstudio/blob/master/js/mviewerstudio.js#L155), it seems that Neogeo's backend returns something like
```
{"firstName": "Jean", "lastName":"Pommier", "userGroups": [{ "slugName":"", "fullName":""}, ...]}
```
and I believe that the flask backend proposed by @RemiDesgrange outputs something like
```
{"firstname": "Jean", "lastname":"Pommier", "username":"jpommier", "org":"myorg", "role":[]}
```

So... snake-case, camelCase of lowercase ? 
Maybe agreeing on the object's content would be nice too, so we don't get lots of custom-backend-specific stuff, wouldn't it ?

**Comment by spelhate** [comment]
I agree with you @jeanpommier it's necessary to define this. I can easily adapt if other developers converge on a proposal.

**Comment by RemiDesgrange** [comment]
You're right @spelhate I'll change to stick to what `user_info.php` provide.

**Comment by bchartier** [comment]
I agree with @RemiDesgrange.
Should we document the backend interface beforehand?

**Comment by spelhate** [comment]
Yes we should. I can document the actual  default  backend. What form ??

**Comment by spelhate** [comment]
I have initialized un first doc to document the (pre)historic php backend. See https://github.com/geobretagne/mviewerstudio/blob/master/srv/README.md and e9467561326becb4760ec7d40a12b4f409bf5ff5

**Comment by bchartier** [comment]
Nice, thank you.
Sorry for not having answered earlier to your question.
I encountered troubles with Github "sign in button" when you posted your question.

**Comment by bchartier** [comment]
I'm working on this issue as we need to merge the pull request for the Camptocamp Python backend (ie https://github.com/geobretagne/mviewerstudio/pull/47).
To sum up:
1. we have a bug identified by @jeanpommier on line 148 of mviewerstudio.js. I have added (in a local branch for the moment) a test to make it work
2. we have some differences on the expected response for the user_info endpoint of the backend as mentioned here: https://github.com/geobretagne/mviewerstudio/tree/master/srv#user_infophp

For this second point what I suggest is to make the code of the various apps (mviewerstudio, its php backend, the Camptocamp Python backend and the Neogeo Python backend) to converge towards the following structure:

```
{
    "user_name": "jdoe",
    "first_name": "John",
    "last_name": "Doe",
    "user_groups": [
        {
           "full_name":  "ACME Company,
           "slug_name":  "acme-company,
           "user_role":  "referent",
           "group_type": "organisation"
        }, {
           "full_name":  "World Company,
           "slug_name":  "world-company,
           "user_role":  "contributor",
           "group_type": "organisation"
        }, {
           "full_name":  "My humble project,
           "slug_name":  "my-humble-project,
           "user_role":  "referent",
           "group_type": "project"
        }
  ]
}
```

The concept of `user_group` is a generalised kind of entity that can represent an organisation, a user group, a project, whatever. The `user_group` slug name is meant to make it possible to store the file produced for a `user_group` in a specific directory on the server.
We could make only the following attribute mandotory: `user_name`.
`user_groups` can be optional but if provided, only the full_name would be mandatory.
The `slug_name` would then be computed by mviewerstudio frontend.
The `full_name` attribute would be treated as the `organisation.legal_name` attribute found in the PHP and Camptocamp Python backends.

Today in Neogeo Python backend, we use the `user_role` attribute in order to manage different credential levels within organisations : simple contributors (contributor) and people acting as administrators for their organisations (referent). One user may have one role for one `user_group` and another role for another `user_group`.

So the structure produced by the PHP backend and the Camptocamp Python backend would be:
```
{
    "first_name": "John",
    "last_name": "Doe",
    "user_groups": [{
        "full_name":  "ACME Company
        }]
}
```

Which would be very close to the original structure and quite simple to implement.

**Comment by bchartier** [comment]
As a first step I suggest to implement in mviewerstudio frontend the capability to deal with either the two structures:
```
{
    "first_name": "John",
    "last_name": "Doe",
    "user_groups": [{
        "full_name":  "ACME Company"
        }]
}
```
```
{
    "first_name": "John",
    "last_name": "Doe",
    "organisation": {"legal_name": "ACME Company"}
}
```

**Comment by bchartier** [comment]
I have juste implemented the support for user_groups and organisation.legal_name in this commit: 55b7566

**Comment by bchartier** [comment]
@RemiDesgrange @jeanpommier @spelhate : should I close this issue?
